### PR TITLE
Builder-dispatch: Generate provenance for built artifacts

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR is a first step into better SLSA-compliance. It makes a few changes to our builder-dispatch workflow:
- generate sha256 hashes for all the artifacts generated (source tarballs and package tarballs) by the build process
- generate intoto provenance attestation for these artifacts using the generic provenance generator (https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/generic) 

At the moment the provenance attestation is only accessible as an action artifact, and is not uploaded along the other artifacts, which is something we might want to do.

The attestation states that the artifacts were generated on a GitHub action runner, and with which exact workflow and parameters.

The attestation can be visualized via `$ cat multiple.intoto.jsonl | jq -r '.payload' | base64 -d | jq`:
```json
{
  "_type": "https://in-toto.io/Statement/v0.1",
  "predicateType": "https://slsa.dev/provenance/v0.2",
  "subject": [
    {
      "name": "./built_pkgs/dnsdist/1.8.0-alpha0.2615.master.g26a500c3a/dnsdist-1.8.0-alpha0.2615.master.g26a500c3a-el-7.tar.bz2",
      "digest": {
        "sha256": "24de79c06d31b5f1ae9e1c94dfeb8320bbd0586e51f178ffa70a788eb3796cb8"
      }
    },
    {
      "name": "./built_pkgs/dnsdist/1.8.0-alpha0.2615.master.g26a500c3a/dnsdist-1.8.0-alpha0.2615.master.g26a500c3a.tar.bz2",
      "digest": {
        "sha256": "1fbe457efc007ffdcd742e6081ee932a46b3e6b8c6a9dc513978755897e4c9ec"
      }
    }
  ],
  "predicate": {
    "builder": {
      "id": "https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@refs/tags/v1.4.0"
    },
    "buildType": "https://github.com/slsa-framework/slsa-github-generator/generic@v1",
    "invocation": {
      "configSource": {
        "uri": "git+https://github.com/rgacogne/pdns@refs/heads/generate-provenance",
        "digest": {
          "sha1": "7d225ddae9979017c1ce1c4bedec4746d57461d8"
        },
        "entryPoint": ".github/workflows/builder-dispatch.yml"
      },
      "parameters": {
        "event_inputs": {
          "is_release": "NO",
          "os": "el-7",
          "product": "dnsdist",
          "ref": "master"
        }
      },
      "environment": {
        "github_actor": "rgacogne",
        "github_actor_id": "1086257",
        "github_base_ref": "",
        "github_event_name": "workflow_dispatch",
        "github_event_payload": {
          "inputs": {
            "is_release": "NO",
            "os": "el-7",
            "product": "dnsdist",
            "ref": "master"
          },
          "ref": "refs/heads/generate-provenance",
          "repository": {
            "allow_forking": true,
            "archive_url": "https://api.github.com/repos/rgacogne/pdns/{archive_format}{/ref}",
            "archived": false,
            "assignees_url": "https://api.github.com/repos/rgacogne/pdns/assignees{/user}",
            "blobs_url": "https://api.github.com/repos/rgacogne/pdns/git/blobs{/sha}",
            "branches_url": "https://api.github.com/repos/rgacogne/pdns/branches{/branch}",
            "clone_url": "https://github.com/rgacogne/pdns.git",
            "collaborators_url": "https://api.github.com/repos/rgacogne/pdns/collaborators{/collaborator}",
            "comments_url": "https://api.github.com/repos/rgacogne/pdns/comments{/number}",
            "commits_url": "https://api.github.com/repos/rgacogne/pdns/commits{/sha}",
            "compare_url": "https://api.github.com/repos/rgacogne/pdns/compare/{base}...{head}",
            "contents_url": "https://api.github.com/repos/rgacogne/pdns/contents/{+path}",
            "contributors_url": "https://api.github.com/repos/rgacogne/pdns/contributors",
            "created_at": "2015-11-03T10:29:53Z",
            "default_branch": "master",
            "deployments_url": "https://api.github.com/repos/rgacogne/pdns/deployments",
            "description": "PowerDNS",
            "disabled": false,
            "downloads_url": "https://api.github.com/repos/rgacogne/pdns/downloads",
            "events_url": "https://api.github.com/repos/rgacogne/pdns/events",
            "fork": true,
            "forks": 1,
            "forks_count": 1,
            "forks_url": "https://api.github.com/repos/rgacogne/pdns/forks",
            "full_name": "rgacogne/pdns",
            "git_commits_url": "https://api.github.com/repos/rgacogne/pdns/git/commits{/sha}",
            "git_refs_url": "https://api.github.com/repos/rgacogne/pdns/git/refs{/sha}",
            "git_tags_url": "https://api.github.com/repos/rgacogne/pdns/git/tags{/sha}",
            "git_url": "git://github.com/rgacogne/pdns.git",
            "has_discussions": false,
            "has_downloads": true,
            "has_issues": false,
            "has_pages": false,
            "has_projects": true,
            "has_wiki": true,
            "homepage": null,
            "hooks_url": "https://api.github.com/repos/rgacogne/pdns/hooks",
            "html_url": "https://github.com/rgacogne/pdns",
            "id": 45459221,
            "is_template": false,
            "issue_comment_url": "https://api.github.com/repos/rgacogne/pdns/issues/comments{/number}",
            "issue_events_url": "https://api.github.com/repos/rgacogne/pdns/issues/events{/number}",
            "issues_url": "https://api.github.com/repos/rgacogne/pdns/issues{/number}",
            "keys_url": "https://api.github.com/repos/rgacogne/pdns/keys{/key_id}",
            "labels_url": "https://api.github.com/repos/rgacogne/pdns/labels{/name}",
            "language": "C++",
            "languages_url": "https://api.github.com/repos/rgacogne/pdns/languages",
            "license": {
              "key": "gpl-2.0",
              "name": "GNU General Public License v2.0",
              "node_id": "MDc6TGljZW5zZTg=",
              "spdx_id": "GPL-2.0",
              "url": "https://api.github.com/licenses/gpl-2.0"
            },
            "merges_url": "https://api.github.com/repos/rgacogne/pdns/merges",
            "milestones_url": "https://api.github.com/repos/rgacogne/pdns/milestones{/number}",
            "mirror_url": null,
            "name": "pdns",
            "node_id": "MDEwOlJlcG9zaXRvcnk0NTQ1OTIyMQ==",
            "notifications_url": "https://api.github.com/repos/rgacogne/pdns/notifications{?since,all,participating}",
            "open_issues": 0,
            "open_issues_count": 0,
            "owner": {
              "avatar_url": "https://avatars.githubusercontent.com/u/1086257?v=4",
              "events_url": "https://api.github.com/users/rgacogne/events{/privacy}",
              "followers_url": "https://api.github.com/users/rgacogne/followers",
              "following_url": "https://api.github.com/users/rgacogne/following{/other_user}",
              "gists_url": "https://api.github.com/users/rgacogne/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/rgacogne",
              "id": 1086257,
              "login": "rgacogne",
              "node_id": "MDQ6VXNlcjEwODYyNTc=",
              "organizations_url": "https://api.github.com/users/rgacogne/orgs",
              "received_events_url": "https://api.github.com/users/rgacogne/received_events",
              "repos_url": "https://api.github.com/users/rgacogne/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/rgacogne/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/rgacogne/subscriptions",
              "type": "User",
              "url": "https://api.github.com/users/rgacogne"
            },
            "private": false,
            "pulls_url": "https://api.github.com/repos/rgacogne/pdns/pulls{/number}",
            "pushed_at": "2023-02-16T10:27:07Z",
            "releases_url": "https://api.github.com/repos/rgacogne/pdns/releases{/id}",
            "size": 81704,
            "ssh_url": "git@github.com:rgacogne/pdns.git",
            "stargazers_count": 2,
            "stargazers_url": "https://api.github.com/repos/rgacogne/pdns/stargazers",
            "statuses_url": "https://api.github.com/repos/rgacogne/pdns/statuses/{sha}",
            "subscribers_url": "https://api.github.com/repos/rgacogne/pdns/subscribers",
            "subscription_url": "https://api.github.com/repos/rgacogne/pdns/subscription",
            "svn_url": "https://github.com/rgacogne/pdns",
            "tags_url": "https://api.github.com/repos/rgacogne/pdns/tags",
            "teams_url": "https://api.github.com/repos/rgacogne/pdns/teams",
            "topics": [],
            "trees_url": "https://api.github.com/repos/rgacogne/pdns/git/trees{/sha}",
            "updated_at": "2023-01-31T16:40:09Z",
            "url": "https://api.github.com/repos/rgacogne/pdns",
            "visibility": "public",
            "watchers": 2,
            "watchers_count": 2,
            "web_commit_signoff_required": false
          },
          "sender": {
            "avatar_url": "https://avatars.githubusercontent.com/u/1086257?v=4",
            "events_url": "https://api.github.com/users/rgacogne/events{/privacy}",
            "followers_url": "https://api.github.com/users/rgacogne/followers",
            "following_url": "https://api.github.com/users/rgacogne/following{/other_user}",
            "gists_url": "https://api.github.com/users/rgacogne/gists{/gist_id}",
            "gravatar_id": "",
            "html_url": "https://github.com/rgacogne",
            "id": 1086257,
            "login": "rgacogne",
            "node_id": "MDQ6VXNlcjEwODYyNTc=",
            "organizations_url": "https://api.github.com/users/rgacogne/orgs",
            "received_events_url": "https://api.github.com/users/rgacogne/received_events",
            "repos_url": "https://api.github.com/users/rgacogne/repos",
            "site_admin": false,
            "starred_url": "https://api.github.com/users/rgacogne/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/rgacogne/subscriptions",
            "type": "User",
            "url": "https://api.github.com/users/rgacogne"
          },
          "workflow": ".github/workflows/builder-dispatch.yml"
        },
        "github_head_ref": "",
        "github_ref": "refs/heads/generate-provenance",
        "github_ref_type": "branch",
        "github_repository_id": "45459221",
        "github_repository_owner": "rgacogne",
        "github_repository_owner_id": "1086257",
        "github_run_attempt": "1",
        "github_run_id": "4193024379",
        "github_run_number": "28",
        "github_sha1": "7d225ddae9979017c1ce1c4bedec4746d57461d8"
      }
    },
    "metadata": {
      "buildInvocationID": "4193024379-1",
      "completeness": {
        "parameters": true,
        "environment": false,
        "materials": false
      },
      "reproducible": false
    },
    "materials": [
      {
        "uri": "git+https://github.com/rgacogne/pdns@refs/heads/generate-provenance",
        "digest": {
          "sha1": "7d225ddae9979017c1ce1c4bedec4746d57461d8"
        }
      }
    ]
  }
}
```
and the certificate via `cat multiple.intoto.jsonl | jq -r '.signatures[0]["cert"]' | openssl x509 -noout -text`:
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            19:1e:58:dd:de:6a:09:b2:91:14:ee:46:e3:85:3a:c3:e7:82:d3:12
        Signature Algorithm: ecdsa-with-SHA384
        Issuer: O = sigstore.dev, CN = sigstore-intermediate
        Validity
            Not Before: Feb 16 10:47:26 2023 GMT
            Not After : Feb 16 10:57:26 2023 GMT
        Subject:
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:46:64:3c:38:c2:5a:b5:6c:32:27:7f:28:2d:ad:
                    d5:a1:13:73:89:4b:52:ae:03:ec:e6:16:bc:26:ce:
                    01:8d:cd:a9:23:18:c9:4a:97:2f:f4:a6:e9:cc:79:
                    67:ca:06:97:61:ac:4b:81:fc:84:5f:f0:30:2c:bf:
                    a5:98:79:1b:d8
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature
            X509v3 Extended Key Usage:
                Code Signing
            X509v3 Subject Key Identifier:
                7D:89:C9:7B:AC:79:49:82:77:19:8E:04:96:D7:12:91:BB:99:FE:D5
            X509v3 Authority Key Identifier:
                DF:D3:E9:CF:56:24:11:96:F9:A8:D8:E9:28:55:A2:C6:2E:18:64:3F
            X509v3 Subject Alternative Name: critical
                URI:https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@refs/tags/v1.4.0
            1.3.6.1.4.1.57264.1.1:
                https://token.actions.githubusercontent.com
            1.3.6.1.4.1.57264.1.2:
                workflow_dispatch
            1.3.6.1.4.1.57264.1.3:
                7d225ddae9979017c1ce1c4bedec4746d57461d8
            1.3.6.1.4.1.57264.1.4:
                Trigger specific package build
            1.3.6.1.4.1.57264.1.5:
                rgacogne/pdns
            1.3.6.1.4.1.57264.1.6:
                refs/heads/generate-provenance
            CT Precertificate SCTs:
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : DD:3D:30:6A:C6:C7:11:32:63:19:1E:1C:99:67:37:02:
                                A2:4A:5E:B8:DE:3C:AD:FF:87:8A:72:80:2F:29:EE:8E
                    Timestamp : Feb 16 10:47:26.733 2023 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                30:45:02:20:38:65:21:06:CA:45:A7:3D:A8:4E:88:80:
                                41:18:CA:7A:83:08:51:22:AD:47:64:FA:97:0D:6A:52:
                                85:A4:2B:97:02:21:00:CA:13:CE:F3:F6:A0:3F:20:1C:
                                F2:D8:73:DB:DA:B8:1B:BD:54:92:1A:EA:D1:21:CF:0B:
                                67:B6:E8:0C:E4:B5:A7
    Signature Algorithm: ecdsa-with-SHA384
    Signature Value:
        30:66:02:31:00:a6:5a:35:cd:8b:01:a7:9e:1a:27:25:f1:d8:
        5c:09:b5:be:c2:a5:2a:cb:92:14:44:57:65:5a:0c:29:4b:17:
        03:c6:c0:41:6c:30:00:30:d2:41:ca:a1:0b:d1:8f:a7:e9:02:
        31:00:82:6f:c7:c9:f7:56:c5:3c:d8:9c:d8:e7:43:41:83:c2:
        a7:62:d3:f2:b5:fb:6f:ab:d9:fc:ec:29:31:ed:ed:f9:d1:89:
        1b:2a:9f:55:7c:ee:75:bf:b6:a9:1c:1d:f7:ae
```

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by Habbie** [comment]
this replaces a big part of #10989 I think?

**Comment by rgacogne** [comment]
> this replaces a big part of #10989 I think?

It aims to, yes, although I need to make sure I did not forget a very important step (or two) :-)

**Comment by rgacogne** [comment]
Things to consider:
- would it make sense to extract packages from the tarballs instead of signing the tarball? It would provide attestation for the packages instead which would be better (actually, why are we generating a tarball of packages, I genuinely cannot remember?)
- separate attestations for tarball and perhaps for each package?
- it would be nice to output the list of packages used during the build and their versions, and to include that in the attestation if possible
- pinning dependencies whenever possible would make sense. @wojas mentioned to me that the version of `pdns-builder` is already tied to the version of the repository present in the attestation which is nice.

**Comment by rgacogne** [comment]
> Things to consider:
> 
>     * would it make sense to extract packages from the tarballs instead of signing the tarball? It would provide attestation for the packages instead which would be better (actually, why are we generating a tarball of packages, I genuinely cannot remember?)
> 
>     * separate attestations for tarball and perhaps for each package?
> 
>     * it would be nice to output the list of packages used during the build and their versions, and to include that in the attestation if possible
> 
>     * pinning dependencies whenever possible would make sense. @wojas mentioned to me that the version of `pdns-builder` is already tied to the version of the repository present in the attestation which is nice.
> 
>     * publishing the attestations along with the source tarball and packages would make sense

It turns out that @pieterlexis got us covered, or at least provided very nice pieces of this puzzle, in `builder/helpers/generate-deb-provenance.py`, `generate-dnf-provenance.py` and `generate-yum-provenance.py` :heart_eyes:

**Comment by rgacogne** [comment]
>     * would it make sense to extract packages from the tarballs instead of signing the tarball? It would provide attestation for the packages instead which would be better (actually, why are we generating a tarball of packages, I genuinely cannot remember?)

Done, the packages are now extracted.

>     * separate attestations for tarball and perhaps for each package?

Done, we now have one attestation for the source tarball, also including the hash of the packages installed on the target OS used to generate the tarball, plus one attestation per target OS.

>     * it would be nice to output the list of packages used during the build and their versions, and to include that in the attestation if possible

Done, one JSON file is now included in each target OS tarball, containing the list of all installed packages with their versions and hashes (courtesy of Pieter Lexis, to be clear).

>     * pinning dependencies whenever possible would make sense. @wojas mentioned to me that the version of `pdns-builder` is already tied to the version of the repository present in the attestation which is nice.

Not done at the moment, but can be done independently.

>     * publishing the attestations along with the source tarball and packages would make sense

Not done, as it might only make sense for actual releases. Perhaps we can automate that if they set up automatic building when a tag is pushed.

**Comment by rgacogne** [comment]
Quick check based on https://slsa.dev/spec/v0.1/requirements

Source requirements
===================

- Version controlled: OK
- Verified history: OK
- Retained indefinitely: OK
- Two-person reviewed: **NOK** (only the reviewer/merger is a trusted person. We would need to have two different reviewers to go to L4)

Build requirements
==================

- Scripted build: OK
- Build service: OK
- Build as code: OK
- Ephemeral environment: OK
- Isolated: **NOK** (we might need to stop using caching for releases to go to L3 and L4)
- Parameterless: **NOK** (we might need to build automatically when a tag is pushed without any external parameter to reach L4)
- Hermetic: **NOK** (we would need to pin everything we install to go to L4)
- Reproducible: OK

Provenance requirements
=======================

- Available: OK
- Authenticated: OK
- Service generated: OK
- Non-falsifiable: OK
- Dependencies complete: **NOK** (we do record system packages with their version and hash, but we would likely need to do the same for python dependencies (or pin them) to reach L4)

- Identifies artifact: OK
- Identifies builder: OK
- Identifies build instructions: OK
- Identifies source code: OK
- Identifies entry point: OK
- Includes all build parameters: **NOK** (see Parameterless)
- Includes all transitive dependencies: **NOK** (see Dependencies complete)
- Includes reproducible info: OK
- Includes metadata: OK

If I'm not mistaken, fixing the `Isolated` requirement would get us to L3. L4 would require a bit more work, and the two-persons reviewed requirement would be quite some hassle.

**Comment by rgacogne** [comment]
- Rebased on master
- Added a reusable workflow to build packages when a tag is pushed
- Refactored builder-dispatch to use the reusable workflow as well

**Comment by rgacogne** [comment]
With the new changes, if I'm not mistaken:

Source requirements
===================

- Version controlled: OK
- Verified history: OK
- Retained indefinitely: OK
- Two-person reviewed: **NOK** (only the reviewer/merger is a trusted person. We would need to have two different reviewers to go to L4)

Build requirements
==================

- Scripted build: OK
- Build service: OK
- Build as code: OK
- Ephemeral environment: OK
- Isolated: OK (we do not use caching when building packages)
- Parameterless: OK (we do not take parameters when we build from a tag)
- Hermetic: **NOK** (we would need to pin everything we install to go to L4)
- Reproducible: OK

Provenance requirements
=======================

- Available: OK
- Authenticated: OK
- Service generated: OK
- Non-falsifiable: OK
- Dependencies complete: **NOK** (we do record system packages with their version and hash, but we would likely need to do the same for python dependencies (or pin them) to reach L4)

- Identifies artifact: OK
- Identifies builder: OK
- Identifies build instructions: OK
- Identifies source code: OK
- Identifies entry point: OK
- Includes all build parameters: OK
- Includes all transitive dependencies: **NOK** (see Dependencies complete)
- Includes reproducible info: OK
- Includes metadata: OK

File: .github/workflows/builder-dispatch.yml [hunk 1]
```diff
@@ -60,6 +60,17 @@ jobs:
...
+    outputs:
+      version: ${{ steps.getversion.outputs.version }}
+      pkghashes-el-7: ${{ steps.pkghashes.outputs.pkghashes-el-7 }}
+      pkghashes-el-8: ${{ steps.pkghashes.outputs.pkghashes-el-8 }}
+      pkghashes-el-9: ${{ steps.pkghashes.outputs.pkghashes-el-9 }}
+      pkghashes-debian-buster: ${{ steps.pkghashes.outputs.pkghashes-debian-buster }}
+      pkghashes-debian-bullseye: ${{ steps.pkghashes.outputs.pkghashes-debian-bullseye }}
+      pkghashes-ubuntu-bionic: ${{ steps.pkghashes.outputs.pkghashes-ubuntu-bionic }}
+      pkghashes-ubuntu-focal: ${{ steps.pkghashes.outputs.pkghashes-ubuntu-focal }}
+      pkghashes-ubuntu-jammy: ${{ steps.pkghashes.outputs.pkghashes-ubuntu-jammy }}
```
Comment (Habbie) [inline .github/workflows/builder-dispatch.yml:72]: I was going to say this list a bit inconvenient, but if we want to be parameterless this list is static anyway - just happens to appear twice in this file, which is workable

**Comment by Habbie** [comment]
> Parameterless: OK (we do not take parameters when we build from a tag)

tag pushes don't cause builds yet, but they could (Pieter's PR has the yaml for it)



> Reproducible: OK

I vaguely recall that sdist might cause today's date to end up in manual pages, which is not terrible, but does mean a difference.

**Review by Habbie — CHANGES_REQUESTED** [review]
nice work! the OS list will need updating before merge, and there are next steps (build on tag push is the most important one I think), but I think we can merge this

**Comment by rgacogne** [comment]
And we should now properly upload the provenance artifacts as well. It's a bit hard for me to actually test that part, though :-/

**Comment by rgacogne** [comment]
I found a way to test most of it, please do not review just yet :)

**Comment by rgacogne** [comment]
@Habbie I did not manage to test everything, especially the part involving the per-repository secrets, but I did test most of the other parts now and fixed a few bugs, I believe this is ready for review!

**Review by Habbie — APPROVED** [review]
great refactor, great work!

**Pull request merged**